Pending price impact

dmax10 opened the pull request:
Store open position price impact on position open instead of charging it directly, charge the total price impact (open + close) on position close
- This allows total price impact to be capped, with the excess being claimable

dmax10 committed:
Store open position price impact on position open instead of charging it directly, charge the total price impact (open + close) on position close

This allows total price impact to be capped, with the excess being claimable

xvi10 commented on contracts/position/PositionEventUtils.sol:
```diff
@@ -59,7 +59,7 @@ library PositionEventUtils {
         eventData.uintItems.setItem(11, "collateralTokenPrice.min", params.collateralTokenPrice.min);
         eventData.uintItems.setItem(12, "sizeDeltaUsd", params.sizeDeltaUsd);
-        eventData.uintItems.setItem(13, "sizeDeltaInTokens", params.sizeDeltaInTokens);
+        eventData.uintItems.setItem(13, "baseSizeDeltaInTokens", params.baseSizeDeltaInTokens);
```
i think this part should not be changed, the event keys are used for e.g. analytics, so we should avoid changing the naming here if possible

dmax10 replied:
Changed back the event key

xvi10 commented on contracts/position/IncreasePositionUtils.sol:
```diff
@@ -99,7 +99,7 @@ library IncreasePositionUtils {
         }
 
-        (cache.priceImpactUsd, cache.priceImpactAmount, cache.sizeDeltaInTokens, cache.executionPrice) = PositionUtils.getExecutionPriceForIncrease(params, prices.indexTokenPrice);
+        (cache.priceImpactUsd, cache.priceImpactAmount, cache.baseSizeDeltaInTokens, cache.executionPrice) = PositionUtils.getExecutionPriceForIncrease(params, prices.indexTokenPrice);
```
maybe we should keep the naming as `sizeDeltaInTokens` instead of changing it to `baseSizeDeltaInTokens`

dmax10 replied:
Changed back to `sizeDeltaInTokens` for the event key, but kept `baseSizeDeltaInTokens` elsewhere as discussed.

xvi10 commented on contracts/position/PositionUtils.sol:
```diff
@@ -710,14 +701,14 @@ library PositionUtils {
         );
 
-        return (priceImpactUsd, priceImpactAmount, sizeDeltaInTokens.toUint256(), executionPrice);
+        return (priceImpactUsd, priceImpactAmount, baseSizeDeltaInTokens, executionPrice);
```
think we can keep the name as `sizeDeltaInTokens` instead of changing it to `baseSizeDeltaInTokens`

dmax10 replied:
Changed back to `sizeDeltaInTokens` for the event key, but kept `baseSizeDeltaInTokens` elsewhere as discussed.

dmax10 committed:
Keep the original event key since it is used for e.g. analytics

dmax10 committed:
Remove comments

dmax10 committed:
Track impact pending amount using a hash from account and market.

Using the positionKey which is the hash of account, market, collateralToken, and isPositive would make it harder to track the impact for the same position on increase/decrease.

dmax10 committed:
test: export functions to generate the impact pending key

dmax10 committed:
test: wip with comments

xdev10 committed:
add @layerzerolabs/lz-evm-protocol-v2

dmax10 committed:
Remove combined key logic for storing the pending impact

dmax10 committed:
Use position store instead of the data store for storing/retrieving position impact pending amount

dmax10 committed:
Fix proportional impact pending amount on decrease

dmax10 committed:
Update "uncapped price impact" and "negative price impact, negative pnl" test cases

dmax10 committed:
wip "capped price impact"

dmax10 committed:
update comments

dmax10 committed:
test update

dmax10 committed:
wip PairMarket "price impact"

dmax10 committed:
Remove blank lines

dmax10 committed:
Update comments

dmax10 committed:
Fix claimable amount logic

dmax10 committed:
comments update

dmax10 committed:
Get the negative impact factor for  position when calcualing the diff

dmax10 committed:
Cap priceImpactUsd to the min negative value

dmax10 committed:
test "Positive & negative impact fees for positions"

dmax10 committed:
Remove impact pending value from data store when position is closed

dmax10 committed:
update test "negative price impact, positive pnl"

dmax10 committed:
Update pnl test cases

dmax10 committed:
Update tests

dmax10 committed:
update test "price impact pair market"

dmax10 committed:
Update life cycle tests

dmax10 committed:
update test and comments

dmax10 committed:
test liquidatable position on collateral decrease

dmax10 committed:
comments update

xvi10 commented on contracts/position/DecreasePositionCollateralUtils.sol:
```diff
@@ -47,6 +49,11 @@ library DecreasePositionCollateralUtils {
         uint256 remainingCostUsd;
     }
 
+    struct PriceImpact {
```
combine into ProcessCollateralCache

dmax10 replied:
Removed PriceImpact struct and added it's fields directly into the ProcessCollateralCache

xvi10 commented on contracts/position/DecreasePositionCollateralUtils.sol:
```diff
@@ -693,4 +740,19 @@ library DecreasePositionCollateralUtils {
+        Price.Props memory indexTokenPrice
+    ) private pure returns (int256, int256) {
+        int256 proportionalImpactPendingAmount = Precision.mulDiv(positionImpactPendingAmount, sizeDeltaUsd, sizeInUsd);
+
```
could add a comment here for why indexTokenPrice.min and indexTokenPrice.max is used for the different cases

dmax10 replied:
Added comment

xvi10 commented on contracts/position/DecreasePositionCollateralUtils.sol:
```diff
@@ -136,9 +144,48 @@ library DecreasePositionCollateralUtils {
+            cache.prices.indexTokenPrice
+        );
+
+        if (values.priceImpactUsd + collateralCache.priceImpact.proportionalImpactPendingUsd < 0) {
```
should add a values.totalPriceImpactUsd = values.priceImpactUsd + collateralCache.priceImpact.proportionalImpactPendingUsd

dmax10 replied:
Added.

xvi10 commented on contracts/position/DecreasePositionCollateralUtils.sol:
```diff
@@ -136,9 +144,48 @@ library DecreasePositionCollateralUtils {
+                values.priceImpactUsd = minPriceImpactUsd;
+            }
+        }
+
```
add values.proportionalImpactPendingUsd to emitPositionDecrease event, the event item key should just be "proportionalImpactPendingUsd" and not "values.proportionalImpactPendingUsd"

dmax10 replied:
Added.

xvi10 commented on contracts/position/DecreasePositionCollateralUtils.sol:
```diff
@@ -136,9 +144,48 @@ library DecreasePositionCollateralUtils {
+            params.order.sizeDeltaUsd()
+        );
+
+        if (collateralCache.priceImpact.cappedTotalImpactUsd > 0) {
```
add a comment that this is for the case where price impact is positive, use totalPriceImpactUsd, the value should be the same as cappedTotalImpactUsd

xvi10 commented on contracts/market/MarketUtils.sol:
```diff
@@ -792,39 +792,42 @@ library MarketUtils {
 
-        if (priceImpactUsd > maxPriceImpactUsdBasedOnMaxPriceImpactFactor) {
-            priceImpactUsd = maxPriceImpactUsdBasedOnMaxPriceImpactFactor;
+        if (totalPriceImpactUsd > maxPriceImpactUsdBasedOnMaxPriceImpactFactor) {
```
add a comment that this is capping for positive price impact

dmax10 replied:
Added

dmax10 committed:
Combine PriceImpact struct into ProcessCollateralCache, add comment why indexTokenPrice.min and indexTokenPrice.max when calculating the proportional impact pending

dmax10 committed:
Add proportionalImpactPendingUsd to emitPositionDecrease event

dmax10 committed:
Split getCappedPositionImpactUsd into capPositiveImpactUsdByPositionImpactPool and capPositiveImpactUsdByMaxPositionImpact

dmax10 committed:
Update price impact capping as follows:

- before calculating execution price for increase and decrease, positive priceImpactUsd is capped by impact pool amount and max impact factor
- on decrease, positive totalImpactUsd is capped by impact pool amount
- on decrease, negative totalImpactUsd is capped by max impact factor, difference is claimable

dmax10 committed:
Rename totalPriceImpactUsd to priceImpactUsd in the capping method

xvi10 commented on contracts/position/DecreasePositionCollateralUtils.sol:
```diff
@@ -136,9 +140,48 @@ library DecreasePositionCollateralUtils {
+        }
+
+        // cap the positive totalImpactUsd by the available amount in the position impact pool
+        collateralCache.cappedTotalImpactUsd = MarketUtils.capPositiveImpactUsdByPositionImpactPool(
```
can we use totalImpactUsd here?

`collateralCache.totalImpactUsd = MarketUtils.capPositiveImpactUsdByPositionImpactPool(`

then we don't need to have a `collateralCache.cappedTotalImpactUsd`, or do they need to be separate?

dmax10 replied:
Sure, removed collateralCache.cappedTotalImpactUsd.

dmax10 committed:
Remove cappedTotalImpactUsd and update totalImpactUsd instead.

dmax10 committed:
Update tests for the new price impact capping logic

dmax10 committed:
update test values

dmax10 committed:
update comments

dmax10 committed:
update test and comments

dmax10 committed:
Update comments
